src/scripting.c

```diff
@@ -128,139 +133,139 @@ void sha1hex(char *digest, char *script, size_t len) {
-    }
+
+static void redisProtocolToLuaType_Bulk(void* ctx, const char* str, size_t len, const char* proto, size_t proto_len) {
+    UNUSED(proto);
+    UNUSED(proto_len);
+    lua_State *lua = ctx;
+    lua_pushlstring(lua,str,len);
 }
 
-char *redisProtocolToLuaType_Status(lua_State *lua, char *reply) {
-    char *p = strchr(reply+1,'\r');
+static void redisProtocolToLuaType_Status(void* ctx, const char* str, size_t len, const char* proto, size_t proto_len) {
+    UNUSED(proto);
+    UNUSED(proto_len);
+    lua_State *lua = ctx;
 
     lua_newtable(lua);
     lua_pushstring(lua,"ok");
-    lua_pushlstring(lua,reply+1,p-reply-1);
+    lua_pushlstring(lua,str,len);
     lua_settable(lua,-3);
-    return p+2;
 }
 
-char *redisProtocolToLuaType_Error(lua_State *lua, char *reply) {
-    char *p = strchr(reply+1,'\r');
+static void redisProtocolToLuaType_Error(void* ctx, const char* str, size_t len, const char* proto, size_t proto_len) {
+    UNUSED(proto);
+    UNUSED(proto_len);
+    lua_State *lua = ctx;
 
     lua_newtable(lua);
     lua_pushstring(lua,"err");
-    lua_pushlstring(lua,reply+1,p-reply-1);
+    lua_pushlstring(lua,str,len);
     lua_settable(lua,-3);
-    return p+2;
 }
 
-char *redisProtocolToLuaType_Aggregate(lua_State *lua, char *reply, int atype) {
-    char *p = strchr(reply+1,'\r');
-    long long mbulklen;
-    int j = 0;
+static void redisProtocolToLuaType_Map(struct ReplyParser* parser, void* ctx, size_t len, const char* proto) {
+    UNUSED(proto);
+    lua_State *lua = ctx;
+    lua_newtable(lua);
+    lua_pushstring(lua, "map");
+    lua_newtable(lua);
+    for (size_t j = 0; j < len; j++) {
+        parseReply(parser,lua);
+        parseReply(parser,lua);
+        lua_settable(lua,-3);
+    }
+    lua_settable(lua,-3);
+}
 
-    string2ll(reply+1,p-reply-1,&mbulklen);
-    if (server.lua_client->resp == 2 || atype == '*') {
-        p += 2;
-        if (mbulklen == -1) {
-            lua_pushboolean(lua,0);
-            return p;
-        }
-        lua_newtable(lua);
-        for (j = 0; j < mbulklen; j++) {
-            lua_pushnumber(lua,j+1);
-            p = redisProtocolToLuaType(lua,p);
-            lua_settable(lua,-3);
-        }
-    } else if (server.lua_client->resp == 3) {
-        /* Here we handle only Set and Map replies in RESP3 mode, since arrays
-         * follow the above RESP2 code path. Note that those are represented
-         * as a table with the "map" or "set" field populated with the actual
-         * table representing the set or the map type. */
-        p += 2;
-        lua_newtable(lua);
-        lua_pushstring(lua,atype == '%' ? "map" : "set");
-        lua_newtable(lua);
-        for (j = 0; j < mbulklen; j++) {
-            p = redisProtocolToLuaType(lua,p);
-            if (atype == '%') {
-                p = redisProtocolToLuaType(lua,p);
-            } else {
-                lua_pushboolean(lua,1);
-            }
-            lua_settable(lua,-3);
-        }
```

Comment:
in resp2 mode, the old code would have kept representing sets and maps as arrays.
1. where is this handled in the new code?
2. do we have test coverage for this, if not, let's add.

Reply:
This is a good point, but I do not think this is even possible, if resp2 is set on the client then the reply will be written as resp2 from the first place. Am I missing something?

Reply:
maybe you're missing `redis.setresp()`? i.e the lua script is the one who controls the response it gets from the lua_client, not the real client that executed Lua.

Reply:
Conclusion of discussion with @oranagra 

`redis.setresp` is setting the protocol on the Lua client (`server.lua_client->resp`). This means that if the user performs `redis.setresp(2)` (or uses it by default), when a command will be executed using `redis.call` it will be written to the Lua client buffer as resp2. This means that checking the protocol number of the client when parsing is redundant.